File: LAB_10/Strategy.java
import java.util.ArrayList;
import java.util.Collections;

interface SortingStrategy {
    void sort(ArrayList<Integer> data);
}

class AscendingSort implements SortingStrategy {
    @Override
    public void sort(ArrayList<Integer> data) {
        Collections.sort(data);
        System.out.println("Sorted in ascending order: " + data);
    }
}

class DescendingSort implements SortingStrategy {
    @Override
    public void sort(ArrayList<Integer> data) {
        Collections.sort(data, Collections.reverseOrder());
        System.out.println("Sorted in descending order: " + data);
    }
}

class Sorter {
    private SortingStrategy strategy;

    public void setStrategy(SortingStrategy strategy) {
        this.strategy = strategy;
    }

    public void sort(ArrayList<Integer> data) {
        if (strategy == null) {
            System.out.println("No sorting strategy set");
            return;
        }
        strategy.sort(data);
    }
}

public class Strategy {
    public static void main(String[] args) {
        ArrayList<Integer> data = new ArrayList<>();
        data.add(3);
        data.add(6);
        data.add(2);
        data.add(8);
        data.add(1);
        data.add(7);

        Sorter sorter = new Sorter();

        sorter.setStrategy(new AscendingSort());
        sorter.sort(data);

        sorter.setStrategy(new DescendingSort());
        sorter.sort(data);
    }
}

File: LAB_10/Strategy.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <functional>


class SortingStrategy {
    public:
        virtual void sort(std::vector<int>& data) = 0;
};

class AscendingSort : public SortingStrategy {
    public:
        void sort(std::vector<int>& data) override {
            std::sort(data.begin(), data.end());
            std::cout << "Sorted in descending order: [";
            for (int i : data) {
                std::cout << i << ", ";
            }
            std::cout << "]" << std::endl;
        }
};

class DescendingSort : public SortingStrategy {
    public:
        void sort(std::vector<int>& data) override {
            std::sort(data.begin(), data.end(), std::greater<int>());
            std::cout << "Sorted in descending order: [";
            for (int i : data) {
                std::cout << i << ", ";
            }
            std::cout << "]" << std::endl;
        }
};

class Sorter {
    private:
        SortingStrategy* sortingStrategy;
    public:
        void setStrategy(SortingStrategy* sortingStrategy) {
            this->sortingStrategy = sortingStrategy;
        }

        void sort(std::vector<int>& data) {
            sortingStrategy->sort(data);
        }
};

int main() {
    std::vector<int>* data = new std::vector<int>();
    data->push_back(4);
    data->push_back(1);
    data->push_back(9);
    data->push_back(3);
    data->push_back(2);

    Sorter *sorter = new Sorter();

    sorter->setStrategy(new AscendingSort());
    sorter->sort(*data);

    sorter->setStrategy(new DescendingSort());
    sorter->sort(*data);

    delete data;
    delete sorter;
}